src/hooks/prerun.ts

```diff
@@ -0,0 +1,28 @@
+import chalk from 'chalk';
+import fs from 'fs-extra';
+import path from 'path';
+import { CREDENTIAL_PREFIX } from '../app-config/auth';
+import AppConfig from '../app-config/config';
+import CredentialManager from '../app-config/credentials';
+import ARCHITECTPATHS from '../paths';
+
+export default async (options: any) => {
+  if (options.Command.id !== 'login') {
```

Comment:
Will this prompt you to login for every cmd except login?

Reply:
Yeah, although I should probably add a local deploy to the list of exemptions. I wanted to add something like this because so many of our commands just do really weird things when you aren't logged in, and there's no indication of what to do or why something gives strange results

Reply:
@davidthor thoughts? maybe each command should dictate whether login is required.

Reply:
I had the same comment, but forgot to file my review haha. Why don't we just catch the 401 properly and handle it accordingly? That way we don't need any logic in the CLI telling us what's allowed and what isn't.

Reply:
I generally like that, but then we also need it for docker and I'm not sure we can determine that without scraping stderr.

Reply:
Perhaps we do both then. Have the handler trigger by default on a 401, but also skip the command entirely if the command has an `auth_required` flag.

Reply:
My thought here is that it's the most configurable if we just add a `@login_required` decorator on functions inside of commands where we need auth. We can't really add it to commands themselves in the case of the deploy command, for example

Reply:
My thought here is that it's the most configurable if we just add a @login_required decorator on functions inside of commands where we need auth. We can't really add it to commands themselves in the case of the deploy command, for example

Reply:
Good point on deploy. I'd still love to keep it as a simple flag. Maybe we could do a flag
`auth_required = true`  for most commands, but then special cases get handled with a getter:
```ts
get auth_required() {
  return !flags.local
}
```

Reply:
I added a function rather than a getter because if I was going to use a getter, a setter would have also been required, and we don't want one

Reply:
Why is a setter required? eslint? Plenty of places in the api where we don't have a setter.

Reply:
Seemed that it had to do with something under the hood for the actual `Command` class. There was also the issue though that we need to parse the deploy command flags, which couldn't be done in a setter